docs/getting-started/part-3/custom-webservice.md

```diff
@@ -0,0 +1,167 @@
+:::
+
+## Get Data Frame
+In this section we will learn how we can avoid sending the requests twice to the API using the execution phase information provided by the smart data lake. We will now implement a simple *if ... else* statement that allows us to simply return a empty data frame with the correct schema in the **Init** phase and to actually query the data in the **Exec** phase. This logic is implemented in the next code snipped and should replace the code currently enclosed between the two `// REPLACE BLOCK` comments.
+```scala
+if(context.phase == ExecutionPhase.Init){
+  // simply return an empty data frame
+  Seq[String]().toDF("responseString")
+    .select(from_json($"responseString", schema.get, Map[String,String]()).as("response"))
+    .select(explode($"response").as("record"))
+    .select("record.*")
+} else {
+  // place the new implementation of currentQueryParameters below this line
+
+  // given the query parameters, generate all requests
+  val departureRequests = currentQueryParameters.map(
+    param => s"${baseUrl}?airport=${param.airport}&begin=${param.begin}&end=${param.end}"
+  )
+  // make requests
+  val departuresResponses = departureRequests.map(request(_))
+  // create dataframe with the correct schema and add created_at column with the current timestamp
+  val departuresDf = departuresResponses.toDF("responseBinary")
+    .withColumn("responseString", byte2String($"responseBinary"))
+    .select(from_json($"responseString", schema.get, Map[String,String]()).as("response"))
+    .select(explode($"response").as("record"))
+    .select("record.*")
+    .withColumn("created_at", current_timestamp())
+  
+  // put simple nextState logic below
+  
+  // return
+   departuresDf
+}
+```
+Don't be confused about some of the comments in the code. They will be of use in the next chapter. If you rebuild the docker image and then restart the program you should see that we do not query the API twice anymore.
+
+## Preserve schema
+
+With this implementation we write the spark data frame of our CustomWebserviceDataObject again in Json format. As a consequence, we loose the schema definition when the Data is read again. To mitigate this behaviour it is beneficial to directly use the `ext-departures` as *inputId* in the `deduplicate-departures` Action. Consequently the first transformer has to be rewritten as well, since the input has changed. Please replace it with the implementation below
+```
+{
+  type = SQLDfTransformer
+  code = "select ext_departures.*, date_format(from_unixtime(firstseen),'yyyyMMdd') dt from ext_departures"
+}
+```
+The old Action `download-departures` can be commented out or deleted.
+
+At the end your config file should look somehting like [this](../config-examples/application-download-part3-custom-webservice.conf) and the Data Object like [this](../config-examples/CustomWebserviceDataObject-1.scala).
```

Comment:
So I guess now I should execute the feed compute-departures, right? When I try to execute it with the provided Files it fails because it cannot connect to the metastore:
`Caused by: ERROR 08001: java.net.UnknownHostException : Error connecting to server metastore on port 1,527 with message metastore.
        at org.apache.derby.client.net.NetAgent.<init>(Unknown Source)
        at org.apache.derby.client.net.NetConnection.newAgent_(Unknown Source)
        at org.apache.derby.client.am.ClientConnection.<init>(Unknown Source)
        at org.apache.derby.client.net.NetConnection.<init>(Unknown Source)
        at org.apache.derby.client.net.ClientJDBCObjectFactoryImpl.newNetConnection(Unknown Source)
        ... 97 more
Caused by: java.net.UnknownHostException: metastore
`
Does this tutorial expect you to have run docker-compose up before so that the metastore is running? I still get this error, even when I have my metastore and polynote running though.

Reply:
If you run it within intellij please comment out the following lines in the application.conf
    #"spark.hadoop.javax.jdo.option.ConnectionURL" = "jdbc:derby://metastore:1527/db;create=true"
    #"spark.hadoop.javax.jdo.option.ConnectionDriverName" = "org.apache.derby.jdbc.ClientDriver"
    #"spark.hadoop.javax.jdo.option.ConnectionUserName" = "sa"
    #"spark.hadoop.javax.jdo.option.ConnectionPassword" = "1234"
And yes, the docker-compose containers should be running, otherwise it will not find the metastore.